lxc/completion.go

```diff
@@ -10,81 +11,204 @@ import (
+		return server.GetNetworkZoneNames()
+	},
+	"profile": func(server lxd.InstanceServer) ([]string, error) {
+		return server.GetProfileNames()
+	},
+	"project": func(server lxd.InstanceServer) ([]string, error) {
+		return server.GetProjectNames()
+	},
+	"storage_pool": func(server lxd.InstanceServer) ([]string, error) {
+		return server.GetStoragePoolNames()
+	},
+	"cluster_member": func(server lxd.InstanceServer) ([]string, error) {
+		return server.GetClusterMemberNames()
+	},
+	"cluster_group": func(server lxd.InstanceServer) ([]string, error) {
+		return server.GetClusterGroupNames()
+	},
 }
 
-// cmpClusterGroups provides shell completion for cluster groups and their remotes.
-// It takes a partial input string and returns a list of matching cluster groups along with a shell completion directive.
-func (g *cmdGlobal) cmpClusterGroups(toComplete string) ([]string, cobra.ShellCompDirective) {
-	cmpDirectives := cobra.ShellCompDirectiveNoFileComp
-
-	resources, _ := g.ParseServers(toComplete)
+var topLevelImageServerResourceNameFuncs = map[string]func(server lxd.ImageServer) ([]string, error){
+	"image": func(server lxd.ImageServer) ([]string, error) {
+		return server.GetImageFingerprints()
+	},
+	"image_alias": func(server lxd.ImageServer) ([]string, error) {
+		return server.GetImageAliasNames()
+	},
+}
 
-	if len(resources) <= 0 {
-		return nil, cobra.ShellCompDirectiveError
+// cmpTopLevelResource is used for general comparison of `<remote>:<resource>` arguments.
+// If no `:` is present in the partial argument `toComplete`, resources from the default remote are returned alongside
+// a list of remote names. The default project for the given remote is used to get the resource list.
+func (g *cmdGlobal) cmpTopLevelResource(entityType string, toComplete string) ([]string, cobra.ShellCompDirective) {
+	remote, partialResourceName, err := g.conf.ParseRemote(toComplete)
+	if err != nil {
+		return handleCompletionError(err)
 	}
 
-	resource := resources[0]
+	names, _ := g.cmpTopLevelResourceInRemote(remote, entityType, partialResourceName)
+	results := make([]string, 0, len(names)+len(g.conf.Remotes))
+	for _, name := range names {
+		var completion string
 
-	cluster, _, err := resource.server.GetCluster()
-	if err != nil || !cluster.Enabled {
-		return nil, cobra.ShellCompDirectiveError
+		if remote == g.conf.DefaultRemote {
+			completion = name
+		} else {
+			completion = remote + ":" + name
+		}
+
+		results = append(results, completion)
 	}
 
-	groups, err := resource.server.GetClusterGroupNames()
-	if err != nil {
-		return nil, cobra.ShellCompDirectiveError
+	directive := cobra.ShellCompDirectiveNoFileComp
+	if !strings.Contains(toComplete, ":") {
+		filters := instanceServerRemoteCompletionFilters(*g.conf)
+		if shared.ValueInSlice(entityType, []string{"image", "image_alias"}) {
+			filters = imageServerRemoteCompletionFilters(*g.conf)
+		}
+
+		remotes, directives := g.cmpRemotes(toComplete, ":", true, filters...)
```

Comment:
```suggestion
		remotes, directives := g.cmpRemotes(toComplete, ":", false, filters...)
```

I don't think we should include a `cobra.ShellCompDirectiveNoSpace` directive for every top level resource. For example, when completing `lxc config set c[TAB]` where `c1` is a running instance, `lxc config set c1 ` should be completed rather than `lxc config set c1`. This allows us to continue tab comleting possible instance keys without needing to enter a `[SPACE]`.

Reply:
Done.